src/aapns/connection.py

```diff
@@ -0,0 +1,440 @@
+    """
+
+    host: str
+    port: int
+    protocol: h2.connection.H2Connection
+    read_stream: Any
+    write_stream: Any
+    should_write: Event
+    channels: Dict[int, Channel] = field(default_factory=dict)
+    reader: Task = field(init=False)
+    writer: Task = field(init=False)
+    closed: bool = False
+    closing: bool = False
+    outcome: Optional[str] = None
+    max_concurrent_streams: int = 100  # recommended in RFC7540#section-6.5.2
+    last_new_sid: int = -1
+    last_sent_sid: int = -1  # client streams are odd
+
+    def __repr__(self):
+        bits = [self.state, f"{self.host}:{self.port}"]
+        if self.state != "closed":
+            bits.extend([f"buffered:{self.buffered}", f"inflight:{self.inflight}"])
+        else:
+            bits.append(f"outcome:{self.outcome}")
+        return "<Connection %s>" % " ".join(bits)
+
+    @classmethod
+    async def create(cls, origin: str, ssl: Optional[SSLContext] = None):
+        url = urlparse(origin)
+        assert url.scheme == "https"
+        assert url.hostname
+        assert not url.username
+        assert not url.password
+        assert not url.path
+        assert not url.params
+        assert not url.query
+        assert not url.fragment
+        host = url.hostname
+        port = url.port or 443
+
+        ssl_context = ssl if ssl else create_ssl_context()
+        assert OP_NO_TLSv1 in ssl_context.options
+        assert OP_NO_TLSv1_1 in ssl_context.options
+        # https://bugs.python.org/issue40111 validate context h2 alpn
+
+        protocol = h2.connection.H2Connection(
+            h2.config.H2Configuration(client_side=True, header_encoding="utf-8")
+        )
+        protocol.local_settings = h2.settings.Settings(
+            client=True,
+            initial_values={
+                # Apple server settings:
+                # HEADER_TABLE_SIZE 4096
+                # MAX_CONCURRENT_STREAMS 1000
+                # INITIAL_WINDOW_SIZE 65535
+                # MAX_FRAME_SIZE 16384
+                # MAX_HEADER_LIST_SIZE 8000
+                h2.settings.SettingCodes.ENABLE_PUSH: 0,
+                h2.settings.SettingCodes.MAX_CONCURRENT_STREAMS: 2 ** 20,
+                h2.settings.SettingCodes.MAX_HEADER_LIST_SIZE: 2 ** 16 - 1,
+                h2.settings.SettingCodes.INITIAL_WINDOW_SIZE: MAX_RESPONSE_SIZE,
+            },
+        )
+
+        protocol.initiate_connection()
+        protocol.increment_flow_control_window(CONNECTION_WINDOW_SIZE)
+
+        try:
+            read_stream, write_stream = await open_connection(
+                host, port, ssl=ssl_context, ssl_handshake_timeout=5
+            )
+            info = write_stream.get_extra_info("ssl_object")
+            # FIXME better exceptions
+            assert info, "HTTP/2 server is required"
+            proto = info.selected_alpn_protocol()
+            assert proto == "h2", "Failed to negotiate HTTP/2"
+        except AssertionError:
+            write_stream.close()
```

Comment:
`write_stream` is not  guaranteed to exist here. the `try` should probably start after `open_connection`

Reply:
done, sorry, I was thinking if I should re-write this and drop asserts...